lib/nerves_livebook/wifi_monitor.ex

```diff
@@ -0,0 +1,123 @@
+
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts \\ []), do: GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+
+  @impl GenServer
+  def init(_opts) do
+    {:ok, %{wlans: []}, {:continue, :check_for_interfaces}}
+  end
+
+  @impl GenServer
+  def handle_continue(:check_for_interfaces, state) do
+    case all_wifi_interfaces() do
+      [] ->
+        # There are no WiFi interfaces on this device that can be configured, so
+        # skip trying to do anything else.
+        {:noreply, state, :hibernate}
+
+      wlans ->
+        {:noreply, %{state | wlans: wlans}, {:continue, :check_connections}}
+    end
+  end
+
+  def handle_continue(:check_connections, state) do
+    if any_connected?(state) do
+      {:noreply, state, :hibenate}
+    else
+      _ = for ifname <- state.wlans, do: VintageNet.subscribe(["interface", ifname, "connection"])
+      {:noreply, state, 3 * 60 * 1000}
+    end
+  end
+
+  @impl GenServer
+  def handle_info(:timeout, state) do
+    _ = for ifname <- state.wlans, do: VintageNet.unsubscribe(["interface", ifname, "connection"])
+
+    _ =
+      unless any_connected?(state) do
+        # TODO: Change some LED here to indicate AP mode?
+        [wlan | _] = Enum.sort(state.wlans)
+        VintageNet.configure(wlan, ap_config(), persist: false)
+      end
+
+    {:noreply, state, :hibernate}
+  end
+
+  def handle_info({VintageNet, ["interface", wlan, "connection"], _old, new, _meta}, state)
+      when new in @connected_status do
+    Logger.info("[WiFiMonitor] #{wlan} in acceptable connection status: #{new}")
+    {:noreply, state, :hibernate}
+  end
+
+  defp all_wifi_interfaces() do
+    for ifname <- VintageNet.all_interfaces(),
+        VintageNet.get(["interface", ifname, "type"]) == VintageNetWiFi or
+          length(VintageNet.get_by_prefix(["interface", ifname, "wifi", :_])) > 0,
+        VintageNet.get(["interface", ifname, "present"]),
+        do: ifname
+  end
+
+  defp any_connected?(%{wlans: wlans}) do
+    Enum.any?(wlans, fn ifname ->
+      VintageNet.get(["interface", ifname, "connection"]) in @connected_status
+    end)
+  end
+
+  defp ap_config() do
+    {:ok, hostname} = :inet.gethostname()
+
+    %{
+      type: VintageNetWiFi,
+      vintage_net_wifi: %{
+        networks: [
+          %{
+            mode: :ap,
+            ssid: to_string(hostname),
+            key_mgmt: :none
+          }
+        ]
+      },
+      ipv4: %{
+        method: :static,
+        address: {192, 168, 0, 1},
+        prefix_length: 24
+      },
+      dhcpd: %{
+        start: {192, 168, 0, 20},
+        end: {192, 168, 0, 254},
+        max_leases: 235,
+        options: %{
+          dns: [{192, 168, 0, 1}],
+          subnet: {255, 255, 255, 0},
+          router: [{192, 168, 0, 1}],
+          domain: "#{hostname}.ap",
+          search: ["#{hostname}.ap"]
+        }
+      },
+      dnsd: %{
+        records: [{"#{hostname}.ap", {192, 168, 0, 1}}]
+      }
+    }
```

Comment:
```suggestion
    {:ok, config} = VintageNetWiFi.Cookbook.open_access_point(to_string(hostname))
    config
```

This isn't quite the same configuration, but I feel like it should be sufficient. The one advantage that I see to yours is the DNS server configuration. I was hoping that mDNS was fine to use here, but if you had troubles with mDNS and this config, I'd change my opinion.

Reply:
I like going with simple first